File: JS/Airport.js
const PassengerPlane = require('./Planes/PassengerPlane');
const MilitaryPlane = require('./Planes/MilitaryPlane');
const MilitaryType = require('./models/militaryType');
const experimentalPlane = require('./Planes/experimentalPlane');

class Airport 
{

    GetPasPl() 
	 {
        let pl = this.planes;
        var x = [];
		
        for (let p of this.planes) 
		{
            if (p instanceof PassengerPlane) {x.push(p);}
        }
		
        return x;
    }

    GetMilitaryPlanes() 
	{
		
        let militaryPlanes = [];
		
        this.planes.forEach(plane => {
			
            if (plane instanceof MilitaryPlane) 
			{
                militaryPlanes.push(plane);
            }
			
        });
		
        return militaryPlanes;
    }

    GetPassengerPlaneWithMaxPassengersCapacity() 
	{
		
        let passengerPlanes = this.GetPasPl();
        let planeWithMaxCapacity = passengerPlanes[0];
		
        for (let i = 0; i < passengerPlanes.length; i++) 
		{
			
            if (passengerPlanes[i].GetPassengersCapacity() > planeWithMaxCapacity.GetPassengersCapacity()) 
			{
                planeWithMaxCapacity = passengerPlanes[i];
            }
			
        }
		
        return planeWithMaxCapacity;
    }

    GetTransportMilitaryPlanes()
	{
		
        let transportMilitaryPlanes = [];
        let militaryPlanes = this.GetMilitaryPlanes();
		
        for (let i = 0; i < militaryPlanes.length; i++) 
		{
			
			if (militaryPlanes[i].GetMilitaryType() == MilitaryType.TYPE_TRANSPORT) 
			{
				transportMilitaryPlanes.push(militaryPlanes[i]);
			}
			
        }
		
        return transportMilitaryPlanes;
    }

    getBomberMilitaryPlanes()
    {
        let bomberMilitaryPlanes = [];
        let militaryPlanes = this.GetMilitaryPlanes();
		
        for (let i = 0; i < militaryPlanes.length; i++) 
		{
			
            if (militaryPlanes[i].GetMilitaryType()=== MilitaryType.BOMBER) 
			{
                bomberMilitaryPlanes.push(militaryPlanes[i]);
            }
			
        }
		
        return bomberMilitaryPlanes;
    }

    constructor(planes)
	{
        this.planes = planes;
    }

    GetExperimentalPlanes() 
	{
        let experimentalPlanes  = [];
		
        this.planes.forEach(plane => 
		{
			
            if (plane instanceof experimentalPlane) 
			{
                experimentalPlanes.push(plane);
            }
			
        });

        return experimentalPlanes;
    }

    SortByMaxDistance()
	{
        this.planes.sort((a, b) => (a.GetMaxFlightDistance() > b.GetMaxFlightDistance()) ? 1 : -1);
        return this;
    }

    /**
     * Sorts by max speed
     * @return Airport
     */
    SortByMaxSpeed() 
	{
        this.planes.sort((a, b) => (a.GetMS() > b.GetMS()) ? 1 : -1);
        return this;
    }

    SortByMaxLoadCapacity() 
	{
        this.planes.sort((a, b) => (a.GetMinLoadCapacity() > b.GetMinLoadCapacity()) ? 1 : -1);
        return this;
    }

    GetPlanes() 
	{
        return this.planes;
    }

    static Print(planes) 
	{
        return JSON.stringify(planes);
    }
}

module.exports = Airport;

File: Java/src/main/java/Airport.java
import Planes.experimentalPlane;
import models.MilitaryType;
import Planes.MilitaryPlane;
import Planes.PassengerPlane;
import Planes.Plane;

import java.util.*;

// version: 1.1
// made by Vitali Shulha
// 4-Jan-2019

public class Airport 
{
    private List<? extends Plane> planes;

    public List<PassengerPlane> GetPasPl() 
    {
        List<? extends Plane> l = this.planes;
        List<PassengerPlane> x = new ArrayList<>();
        
        for (Plane p : l) 
        {
        	
        	if (p instanceof PassengerPlane) 
        	{
        		x.add((PassengerPlane) p);
        	}
        	
        }
        
        return x;
    }

    public List<MilitaryPlane> GetMilitaryPlanes()
    {
        List<MilitaryPlane> militaryPlanes = new ArrayList<>();
        
        for (Plane plane : planes)
        {
        	
            if (plane instanceof MilitaryPlane) 
            {
                militaryPlanes.add((MilitaryPlane) plane);
            } 
            
        } 
        
        return militaryPlanes;
    }

    public PassengerPlane GetPassengerPlaneWithMaxPassengersCapacity() 
    {
        List<PassengerPlane> passengerPlanes = GetPasPl();
        PassengerPlane planeWithMaxCapacity = passengerPlanes.get(0);
        
        for (int i = 0; i < passengerPlanes.size(); i++) 
        {
        	
            if (passengerPlanes.get(i).GetPassengersCapacity() > planeWithMaxCapacity.GetPassengersCapacity()) 
            {
                planeWithMaxCapacity = passengerPlanes.get(i);
            }
            
        }

        return planeWithMaxCapacity;
    }

    public List<MilitaryPlane> GetTransportMilitaryPlanes() 
    {
    	List<MilitaryPlane> transportMilitaryPlanes = new ArrayList<>();
    	List<MilitaryPlane> militaryPlanes = GetMilitaryPlanes();
    	
    	for (int i = 0; i < militaryPlanes.size(); i++) 
    	{
    		MilitaryPlane plane = militaryPlanes.get(i);
    		
    		if (plane.getType() == MilitaryType.TRANSPORT) 
    		{
    			transportMilitaryPlanes.add(plane);
    		}
    	}
    	
    	return transportMilitaryPlanes;
    
    }

    public List<MilitaryPlane> GetBomberMilitaryPlanes() 
    {
        List<MilitaryPlane> bomberMilitaryPlanes = new ArrayList<>();
        List<MilitaryPlane> militaryPlanes = GetMilitaryPlanes();
        
        for (int i = 0; i < militaryPlanes.size(); i++) 
        {
            MilitaryPlane plane = militaryPlanes.get(i);
            
            if (plane.getType() == MilitaryType.BOMBER) 
            {
                bomberMilitaryPlanes.add(plane);
            }
            
        }
        
        return bomberMilitaryPlanes;

    }

    public List<experimentalPlane> GetExperimentalPlanes() 
    {
        List<experimentalPlane> experimentalPlanes = new ArrayList<>();
        
        for (Plane plane : planes) 
        {
        	
            if (plane instanceof experimentalPlane) 
            {
                experimentalPlanes.add((experimentalPlane) plane);
            }
            
        }
        
        return experimentalPlanes;
    }

    public Airport SortByMaxDistance() 
    {
        Collections.sort(planes, new Comparator<Plane>() 
        {
            public int compare(Plane o1, Plane o2) 
            {
                return o1.Get_Max_Flight_Distance() - o2.Get_Max_Flight_Distance();
            }
        });
        
        return this;
    }


    /**
     * Sorts by max speed
     * @return Airport
     */
    public Airport SortByMaxSpeed() 
    {
        Collections.sort(planes, new Comparator<Plane>() 
        {
            public int compare(Plane o1, Plane o2) 
            {
                return o1.getMS() - o2.getMS();
            }
        });
        
        return this;
    }

    public Airport SortByMaxLoadCapacity() 
    {
        Collections.sort(planes, new Comparator<Plane>() 
        {
            public int compare(Plane o1, Plane o2) 
            {
                return o1.GetMinLoadCapacity() - o2.GetMinLoadCapacity();
            }
        });
        
        return this;
    }

    public List<? extends Plane> GetPlanes() 
    {
        return planes;
    }

    private void print(Collection<? extends Plane> collection) 
    {
        Iterator<? extends Plane> iterator = collection.iterator();
        
        while (iterator.hasNext()) 
        {
            Plane plane = iterator.next();
            System.out.println(plane);
        }
    }

    @Override
    public String ToString() 
    {
        return "Airport{" +
                "Planes=" + planes.toString() +
                '}';
    }

    //Constructor
    public Airport(List<? extends Plane> planes)
    {
        this.planes = planes;
    }

}
